libs/wire-api/src/Wire/API/Conversation.hs

```diff
@@ -436,9 +446,47 @@ data AccessRole
   deriving (Arbitrary) via (GenericUniform AccessRole)
   deriving (ToJSON, FromJSON, S.ToSchema) via Schema AccessRole
 
+data AccessRoleV2
+  = TeamMemberAccessRole
+  | GuestAccessRole
+  | ServiceAccessRole
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving (Arbitrary) via (GenericUniform AccessRoleV2)
+  deriving (ToJSON, FromJSON, S.ToSchema) via Schema AccessRoleV2
+
+toAccessRoles :: AccessRole -> Set AccessRoleV2
+toAccessRoles = \case
+  PrivateAccessRole -> Set.fromList []
+  TeamAccessRole -> Set.fromList [TeamMemberAccessRole]
+  ActivatedAccessRole -> Set.fromList [TeamMemberAccessRole, GuestAccessRole]
+  NonActivatedAccessRole -> Set.fromList [TeamMemberAccessRole, GuestAccessRole, ServiceAccessRole]
+
+-- todo(leif): check if this works, might be better to return Maybe and let the call side decide what to do in case of incompatibility
+toAccessRole :: Set AccessRoleV2 -> AccessRole
+toAccessRole accessRoles
+  | Set.null accessRoles = NonActivatedAccessRole
+  | Set.fromList [TeamMemberAccessRole] == accessRoles = TeamAccessRole
+  | Set.fromList [GuestAccessRole] == accessRoles = NonActivatedAccessRole
+  | Set.fromList [ServiceAccessRole] == accessRoles = NonActivatedAccessRole
+  | Set.fromList [TeamMemberAccessRole, GuestAccessRole] == accessRoles = NonActivatedAccessRole
+  | Set.fromList [TeamMemberAccessRole, ServiceAccessRole] == accessRoles = NonActivatedAccessRole
+  | Set.fromList [GuestAccessRole, ServiceAccessRole] == accessRoles = NonActivatedAccessRole
+  | Set.fromList [TeamMemberAccessRole, GuestAccessRole, ServiceAccessRole] == accessRoles = NonActivatedAccessRole
+  | otherwise = ActivatedAccessRole
```

Comment:
It would be really bad if this function couldn't be total, but I think it can be.  (Doesn't the tech spec say how it can?)

if you use a `ViewPattern`, the runtime doesn't have to convert the set as many times as there are cases here, and you can use `case` with its completeness warnings, and expand the `otherwise` case:

```suggestion
toAccessRole :: Set AccessRoleV2 -> AccessRole
toAccessRole accessRoles = case Set.fromList accessRoles of
  [] -> NonActivatedAccessRole
  [TeamMemberAccessRole] -> TeamAccessRole
  [...] -> ...
```